### DISCUSSION TITLE
Custom script on first boot

### DISCUSSION THREAD

**Discussion by mohanjith** [drakkar-lig/debootstick#41] [Q&A] [answered]
Is it possible to provide a custom script to be run on the first boot? 

We have some installation steps that cannot take place in a chroot which could be run on first boot.

**Comment by eduble** [comment, marked as answer]
`debootstick` has no specific option for this, but you can modify your OS tree for this purpose before you call `debootstick`.
An easy way is to add a systemd unit with `ConditionFirstBoot=yes`, see https://www.undrground.org/2021/01/25/adding-a-single-run-task-via-systemd/ for instance. Would that work for you?

**Comment by mohanjith** [comment]
ConditionFirstBoot=yes doesn't work because the /etc/machine-id is populated by debootstrap. 

Ideally I want to time this after debootstick install script finishes. Let's see what workaround I can come up with.

**Comment by mohanjith** [comment]
I created a service of Type=oneshot which waits for getty command to be moved back before doing it's actions. 

It would be cool if debootstick had a means to execute a custom script after performing initial installation steps.

**Reply by eduble** [reply]
I see. This custom code would be called at the end of [scripts/live/init/first-init.sh](https://github.com/drakkar-lig/debootstick/blob/dev/scripts/live/init/first-init.sh) then.
This could be triggered by an option "--run-on-first-boot <path>" handled [in main script](https://github.com/drakkar-lig/debootstick/blob/7957be8b04d15c16577bbd346b71075b7899aad9/debootstick#L13). It would store this optional path as a variable [in /dbstck.conf](https://github.com/drakkar-lig/debootstick/blob/7957be8b04d15c16577bbd346b71075b7899aad9/debootstick#L313).
Then this variable would be available [in script first-init.sh](https://github.com/drakkar-lig/debootstick/blob/7957be8b04d15c16577bbd346b71075b7899aad9/scripts/live/init/first-init.sh#L5) to optionally call this first-boot code.

I have little time these days, but if you are interested feel free to write a pull request.

**Comment by kgilmer** [comment]
FWIW I found there were some quirks in handling user input in such a script.  I desired to collect the installation details upon first boot from the user to make things a bit more standardized.  The fruits of my efforts are here in case it's any help ~ https://github.com/kgilmer/regolith-iso-builder-sandbox/tree/main/.github/scripts.  `systemd` is used and some tty redirects preserve the text console while the system is booting.

**Reply by eduble** [reply]
Hi,
Just in case you were not aware, `debootstick` does provides an option `--run-on-first-boot FIRST_BOOT_SCRIPT_FILE` now:
https://manpages.debian.org/testing/debootstick/debootstick.8.en.html#run
This was not suitable for what you wanted?

**Reply by kgilmer** [reply]
Yes I tried that, but was unable to cleanly collect user input from my script, due to TTY contention I believe
